### PR TITLE
feat: Payment network any-to-eth in the smartcontracts

### PR DESCRIPTION
**Pull request by vrolland**

### CODE REVIEW

**Review by yomarion — APPROVED** [review]
Excellent, well done

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  );
+  
+  // Event to declare a transfer with a reference
+  event TransferWithReferenceAndFee(
+    address to,
+    uint256 amount,
+    bytes indexed paymentReference,
+    uint256 feeAmount,
+    address feeAddress
+  );
```
Comment (benjlevesque) [inline packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: why is this event re-defined, shouldn't it use the eth-fee-proxy contract?

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+   * @notice Performs an ETH transfer with a reference computing the payment amount based on the request amount
+   * @param _to Transfer recipient of the payement
+   * @param _requestAmount Request amount
+   * @param _path Conversion path
+   * @param _paymentReference Reference of the payment related
+   * @param _feeAmount The amount of the payment fee
+   * @param _feeAddress The fee recipient
+   * @param _maxRateTimespan Max time span with the oldestrate, ignored if zero
+   */
+  function transferWithReferenceAndFee(
```
Comment (benjlevesque) [inline packages/smart-contracts/src/contracts/EthConversionProxy.sol:53]: shouldn't this be a call to eth-fee-proxy contract (same as previous comment)

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    bytes calldata _paymentReference,
+    uint256 _feeAmount,
+    address _feeAddress,
+    uint256 _maxRateTimespan
+  )
+    external
+    payable
+  {
+    require(
+      _path[_path.length - 1] == nativeTokenHash,
```
Comment (benjlevesque) [inline packages/smart-contracts/src/contracts/EthConversionProxy.sol:66]: what happens if _path is empty?

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+import "./ChainlinkConversionPath.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+/**
+ * @title EthConversionProxy
+ * @notice This contract convert from chainlink then swaps ETH
+ *         before paying a request thanks to a conversion payment proxy
+ *         the dependance with ReentrancyGuard is required to perform
+ *         "transferExactEthWithReferenceAndFee" of the eth-fee-proxy contract
+  */
```
Comment (benjlevesque) [inline packages/smart-contracts/src/contracts/EthConversionProxy.sol:13]:
```suggestion
/**
 * @title EthConversionProxy
 * @notice This contract converts from chainlink then swaps ETH (or native token)
 *         before paying a request thanks to a conversion payment proxy.
 *         The inheritance from ReentrancyGuard is required to perform
 *         "transferExactEthWithReferenceAndFee" on the eth-fee-proxy contract
  */
```

File: packages/smart-contracts/test/contracts/EthConversionProxy.test.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  let signer: Signer;
+  const smallAmountInFIAT = BigNumber.from('100000000'); 
+  const smallerAmountInFIAT = BigNumber.from('10000000');
+  const referenceExample = '0xaaaa';
+
+  const currencyManager = CurrencyManager.getDefault();
+
+  const ETH_address = currencyManager.fromSymbol('ETH')!.hash;
+  const USD_address = currencyManager.fromSymbol('USD')!.hash;
+  const EUR_address = currencyManager.fromSymbol('EUR')!.hash;
```
Comment (benjlevesque) [inline packages/smart-contracts/test/contracts/EthConversionProxy.test.ts:30]:
```suggestion
  const ETH_hash = currencyManager.fromSymbol('ETH')!.hash;
  const USD_hash = currencyManager.fromSymbol('USD')!.hash;
  const EUR_hash = currencyManager.fromSymbol('EUR')!.hash;
```

File: packages/smart-contracts/test/contracts/EthConversionProxy.test.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+  before(async () => {
+    [from, to, feeAddress] = (await ethers.getSigners()).map((s) => s.address);
+    [signer] = await ethers.getSigners();
+
+    chainlinkPath = chainlinkConversionPath.connect(network.name, signer);
+    ethFeeProxy = await new EthereumFeeProxy__factory(signer).deploy();
+    testEthConversionProxy = await new EthConversionProxy__factory(signer).deploy(
+      ethFeeProxy.address,
+      chainlinkPath.address,
+      ETH_address,
```
Comment (benjlevesque) [inline packages/smart-contracts/test/contracts/EthConversionProxy.test.ts:46]:
```suggestion
      ETH_hash,
```

File: packages/smart-contracts/test/contracts/EthConversionProxy.test.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+use(solidity);
+
+describe('contract: EthConversionProxy', () => {
+  let from: string;
+  let to: string;
+  let feeAddress: string;
+  let signer: Signer;
+  const smallAmountInFIAT = BigNumber.from('100000000'); 
+  const smallerAmountInFIAT = BigNumber.from('10000000');
```
Comment (benjlevesque) [inline packages/smart-contracts/test/contracts/EthConversionProxy.test.ts:23]: why not `amountInFiat` and `feesAmountInFiat`? any reason for the `small` qualification? does it have something to do with the test setup?

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 1 above)
Comment (vrolland) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: as we do a delegate call the event `TransferWithReferenceAndFee()` will be emit by this this contract.
We declare it to have it in the ABI, making the event fliter either.

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 2 above)
Comment (vrolland) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:53]: Yes, there is a delegate call to the `eth-fee-proxy` contract in this function.

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 3 above)
Comment (vrolland) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:66]: `_path.length - 1` will throw (underflow)

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 1 above)
Comment (benjlevesque) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: hum OK, I didn't know this was needed, thanks

Comment (vrolland) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: It's not required but convenient :)

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 2 above)
Comment (benjlevesque) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:53]: OK my bad I was confused by the name, it's ok

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 3 above)
Comment (benjlevesque) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:66]: which is OK I guess... thanks!

File: packages/smart-contracts/src/contracts/EthConversionProxy.sol (hunk 1 above)
Comment (benjlevesque) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: could be useful to add a comment?

Comment (vrolland) [reply packages/smart-contracts/src/contracts/EthConversionProxy.sol:41]: I will then

**Pull request merged**